open-apparel-registry/open-apparel-registry
# Add AWS Batch job retries and failure reporting to Rollbar via AWS Lambda

## Overview

- Set all jobs to retry up to 3 times in an attempt to automatically recover from transient issues that may cause a job to fail.
- Set up job failure notifications by setting up a CloudWatch event to detect failures and connecting that event to a Lambda function that posts the details of the failure to Rollbar. 

Connects #457

## Demo

### Tasks being retried
<img width="801" alt="Screen Shot 2019-05-15 at 1 47 07 PM" src="https://user-images.githubusercontent.com/17363/57808392-1bc00400-7718-11e9-9518-3d5531c641c0.png">

## Job failure reported in Rollbar
<img width="1229" alt="Screen Shot 2019-05-15 at 1 47 31 PM" src="https://user-images.githubusercontent.com/17363/57808408-24b0d580-7718-11e9-846c-488d55ad330b.png">


## Notes

The batch processing management command is designed to only process items in a
specific state so rerunning a partially completed job will not process any items
multiple times.

Posting job failures to Rollbar via a Lambda function has been use successfully in another project.

## Testing Instructions

This must be deployed to staging as batch processing only works in AWS.

There is a testing commit that makes jobs fail 80% of the time that will be dropped prior to merging.

This was tested on staging by manually deploying 71045fd. 

I verify that dependent jobs do not fail until all of the retries of the parent job have also failed.

* Register and confirm an account
* Upload a facility list
  * Verify in the AWS Batch console that failed jobs are retried.
  * Verify that job failures are reported to Rollbar.

## Checklist
